File: frontend/src/lib/stories.ts
import { apiFetch } from "@/lib/api";
import type { SourceItem } from "@/lib/research";

// Mirrors backend app/routers/stories.py : CreateStoryRequest
export interface CreateStoryRequest {
  topic: string;
  language_code?: string;
  sources: SourceItem[];
  pasted_text: string;
}

// Mirrors SourceOut / StoryDetail in app/routers/stories.py
export interface SourceOut {
  id: number;
  type: string;
  title: string | null;
  url: string | null;
  citation_text: string | null;
}

// Mirrors backend app/schemas/brief.py
export interface LevelScopes {
  just_starting: string;
  beginner: string;
  intermediate: string;
  advanced: string;
}

export interface StoryBrief {
  overview: string;
  angle: string;
  spine: string;
  levels: LevelScopes;
}

export interface StoryDetail {
  id: number;
  topic_text: string;
  language_code: string;
  status: string;
  brief: StoryBrief | null;
  sources: SourceOut[];
}

export function createStory(body: CreateStoryRequest): Promise<{ id: number }> {
  return apiFetch<{ id: number }>("/api/stories", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function getStory(id: number): Promise<StoryDetail> {
  return apiFetch<StoryDetail>(`/api/stories/${id}`);
}

// Generate the brief from the story's saved sources (extract + LLM). Slow (~30-60s).
export function generateBrief(id: number): Promise<StoryBrief> {
  return apiFetch<StoryBrief>(`/api/stories/${id}/brief`, { method: "POST" });
}

// Save the creator's edited brief.
export function updateBrief(id: number, brief: StoryBrief): Promise<StoryBrief> {
  return apiFetch<StoryBrief>(`/api/stories/${id}/brief`, {
    method: "PUT",
    body: JSON.stringify(brief),
  });
}

File: backend/app/routers/stories.py
"""Story endpoints: create a draft story from selected sources, and read it back."""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.db import crud
from app.db.session import get_db
from app.schemas.source import SourceItem

router = APIRouter(prefix="/api", tags=["stories"])


class CreateStoryRequest(BaseModel):
    topic: str
    language_code: str = "es-LA"
    sources: list[SourceItem] = []
    pasted_text: str = ""


class CreateStoryResponse(BaseModel):
    id: int


class SourceOut(BaseModel):
    id: int
    type: str
    title: str | None
    url: str | None
    citation_text: str | None


class StoryDetail(BaseModel):
    id: int
    topic_text: str
    language_code: str
    status: str
    brief: dict | None
    sources: list[SourceOut]


@router.post("/stories", response_model=CreateStoryResponse)
def create_story(req: CreateStoryRequest, db: Session = Depends(get_db)) -> CreateStoryResponse:
    """Persist a draft Story with its selected web sources + pasted content."""
    story = crud.create_story(
        db,
        topic=req.topic,
        language_code=req.language_code,
        sources=req.sources,
        pasted_text=req.pasted_text,
    )
    return CreateStoryResponse(id=story.id)


@router.get("/stories/{story_id}", response_model=StoryDetail)
def get_story(story_id: int, db: Session = Depends(get_db)) -> StoryDetail:
    """Return a story with its brief and sources (for reopening a draft)."""
    story = crud.get_story(db, story_id)
    if story is None:
        raise HTTPException(status_code=404, detail="Story not found")
    sources = crud.get_sources(db, story_id)
    return StoryDetail(
        id=story.id,
        topic_text=story.topic_text,
        language_code=story.language_code,
        status=story.status,
        brief=story.brief,
        sources=[
            SourceOut(
                id=s.id,
                type=s.type,
                title=s.title,
                url=s.url,
                citation_text=s.citation_text,
            )
            for s in sources
        ],
    )
